core/server/common/src/main/java/alluxio/master/journal/raft/JournalStateMachine.java

```diff
@@ -682,13 +682,35 @@ public boolean isSnapshotting() {
   public void notifyLeaderChanged(RaftGroupMemberId groupMemberId, RaftPeerId raftPeerId) {
     if (mRaftGroupId == groupMemberId.getGroupId()) {
       mIsLeader = groupMemberId.getPeerId() == raftPeerId;
+      if (!mIsLeader) {
+        // If this state machine wins an election right after starting or right after losing
+        // another election, it will not trigger notifyExtendedNoLeader. This is therefore the
+        // last point of contact with another leader.
+        mJournalSystem.setLastLeaderTime(System.currentTimeMillis());
+      }
       mJournalSystem.notifyLeadershipStateChanged(mIsLeader);
     } else {
       LOG.warn("Received notification for unrecognized group {}, current group is {}",
           groupMemberId.getGroupId(), mRaftGroupId);
     }
   }
 
+  /**
+   * This function is triggered when the state machine becomes CANDIDATE. This function's
+   * activation is regulated by the RaftServerConfigKeys.Notification.setNoLeaderTimeout() call
+   * in RaftJournalSystem.
+   */
+  @Override
+  public void notifyExtendedNoLeader(RaftProtos.RoleInfoProto roleInfoProto) {
+    long lastLeaderElapsedTimeMs = roleInfoProto.getCandidateInfo().getLastLeaderElapsedTimeMs();
+    if (roleInfoProto.getRole() == RaftProtos.RaftPeerRole.CANDIDATE) {
+      LOG.info("Transitioned to CANDIDATE, last leader was {}ms ago", lastLeaderElapsedTimeMs);
+      mJournalSystem.setLastLeaderTime(System.currentTimeMillis() - lastLeaderElapsedTimeMs);
+    } else {
+      LOG.warn("Triggered notifyExtendedNoLeader when role is {}", roleInfoProto.getRole());
```

Comment:
why not set last-leader-time, when the state-machine is say transitioned to PASSIVE or FOLLOWER?

Reply:
I see now how "lastLeaderTime" is confusing, it's not meant to be when this master was leader last, but rather the last time this master interacted with a leading master.
This measurement is meant to describe the time elapsed between when the last leading master lost leadership and when a new leader is up caught up and serving requests. This means measuring the time between when it becomes CANDIDATE and when it becomes LEADER.
This method (`notifyExtendedNoLeader`) is not triggered if the master launches a new election right after losing a previous one. Therefore, I use `notifyLeaderChanged` to mark the last time we saw a leading master in this corner case.